lib/selinux.c

```diff
@@ -98,6 +98,136 @@ int reset_selinux_file_context (void)
 	return 0;
 }
 
+/*
+ *  libselinux uses this callback when access gets denied or other
+ *  events happen. If audit is turned on, messages will be reported
+ *  using audit netlink, otherwise they will be logged using the usual
+ *  channels.
+ *
+ *  Code adopted from systemd, which adopted from dbus.
+ */
+__attribute__((__format__ (printf, 2, 3)))
+static int log_callback (int type, const char *fmt, ...) {
+	va_list ap;
+	char *buf;
+	int r;
+#if HAVE_AUDIT
+	int fd;
+#endif
+
+	va_start (ap, fmt);
+	r = vasprintf (&buf, fmt, ap);
+	va_end (ap);
+
+	if (r < 0) {
+		free (buf);
+		return 0;
+	}
+
+	fprintf (stderr, _("%s: libselinux: %s\n"),
+			    Prog, buf);
+	SYSLOG ((LOG_WARN, "libselinux: %s", buf));
+
+#if HAVE_AUDIT
+	fd = audit_help_open ();
+
+	if (fd >= 0) {
+		uid_t audit_uid = geteuid ();
+
+		if (SELINUX_AVC == type)
+			audit_log_user_avc_message (fd,
+			    AUDIT_USER_AVC, buf, NULL, NULL,
+			    audit_uid, 0);
+		else if (SELINUX_ERROR == type)
+			audit_log_user_avc_message (fd,
+			    AUDIT_USER_SELINUX_ERR, buf, NULL,
+			    NULL, audit_uid, 0);
+	}
+#endif
+
+	free (buf);
+
+	return 0;
+}
+
+/*
+ * check_selinux_permit - Checks whether SELinux grants the given
+ *                        operation
+ *
+ *   Parameter is the SELinux permission name, e.g. rootok
+ *
+ *   Returns 0 when permission is granted
+ *                  or something failed but running in
+ *                  permissive mode
+ */
+int check_selinux_permit (const char *perm_name)
+{
+	return check_selinux_permit_shortcut (perm_name, NULL, 0);
+}
+
+/*
+ * check_selinux_permit - Checks whether SELinux grants the given
+ *                        operation
+ *
+ *   First parameter is the SELinux permission name, e.g. rootok
+ *   Second parameter if not NULL enables a shortcut for changes
+ *       on non-root and matching SELinux user.
+ *   Third parameter tells the shortcut what userid gets changed.
+ *       (if no shortcut is intented, pass 0)
+ *
+ *   Returns 0 when permission is granted
+ *                  or something failed but running in
+ *                  permissive mode
+ */
+int check_selinux_permit_shortcut (const char *perm_name,
+                                      const char *changed_user,
+                                      uid_t changed_uid)
+{
+	char *user_context_str;
+	int r;
+
+	if (0 == is_selinux_enabled ()) {
+		return 0;
+	}
+
+	selinux_set_callback (SELINUX_CB_LOG, (union selinux_callback) log_callback);
+
+	if (getprevcon (&user_context_str) != 0) {
+		fprintf (stderr,
+		    _("%s: can not get previous SELinux process context: %s\n"),
+		    Prog, strerror (errno));
+		SYSLOG ((LOG_WARN,
+		    "can not get previous SELinux process context: %s",
+		    strerror (errno)));
+		freecon (user_context_str);
+		return (security_getenforce () != 0);
+	}
+
+	if (changed_user) {
+		/* get the "user" portion of the context
```

Comment:
I am of the opinion that this entire if statement and its associated dependencies can just go away, i.e. no shortcut logic because SELinux user identity will generally never be the same as Linux username on any modern Linux distro. Worst case is that we'll trigger a passwd permission check that wasn't truly needed.

Reply:
Do you think we need this logic?

Reply:
I agree with removing this logic; I'd like to get an OK from upstream.
Should I announce this change on the mailing-list?

Reply:
Sure, you could ask there. May also want to compare with source code of passwd, which is a separate package in Fedora?

Reply:
Announced
(p.s.: fedora's passwd does not use something like this)